New component OpenPlantbook

Olen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
This PR is a preparation for a later PR to the Plant component.
The Openplantbook component itself does not do much, except providing a way for Home Assistant to communicate with the OpenPlantbook API and online DB to fetch data.
https://open.plantbook.io/

The reason this is added as a separate component is to ensure the Plant component will continue working even if the user does not have a OpenPlantbook account.



<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
The component requires the user to get a client id and secret to the OpenPlantbook API.  It is configured using a config flow.
The Openplantbook API is the exposed to HA as a class where other components (e.g. the Plant component) can fetch data.
A PR for the Plant component where it can take advantage of the OpenPlantbook API will be submitted once this component is accepted.

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Olen committed:
Committing scaffold tests

Olen committed:
Committing scaffold tests

Olen committed:
Adding init

Olen committed:
Adding init

Olen committed:
Config flow

Olen committed:
Config flow

Olen committed:
Constants

Olen committed:
Constants

Olen committed:
Adding myself

Olen committed:
Adding myself

Olen committed:
Adding openplantbook to config_flows

Olen committed:
Adding openplantbook to config_flows

Olen committed:
Englosh strings

Olen committed:
Englosh strings

Olen committed:
Adding manifest

Olen committed:
Adding manifest

Olen committed:
Strings

Olen committed:
Strings

Olen committed:
Fix some translations
Better error handling
General cleanup

Olen committed:
Fix some translations
Better error handling
General cleanup

Olen committed:
Add strings

Olen committed:
Add strings

Olen committed:
Updated strings

Olen committed:
Updated strings

Olen committed:
Delete a comment

Olen committed:
Delete a comment

Olen committed:
Trying to fix tests

Olen committed:
Trying to fix tests

frenck committed:
Upgrade shodan to 1.24.0 (#42199)

Olen committed:
Fix some pylint errors

Olen committed:
Fix some pylint errors

Olen committed:
Fix tests

Olen committed:
Fix tests

Olen committed:
More pylint-fixes

Olen committed:
More pylint-fixes

Olen committed:
Raise from...

Olen committed:
Raise from...

RobBie1221 commented on homeassistant/components/openplantbook/__init__.py:
```diff
@@ -0,0 +1,97 @@
+        }
+        try:
+            session = async_get_clientsession(self.hass)
+            async with session.get(url, headers=headers) as result:
```
Communication to external devices or services must be wrapped in an external Python library hosted on pypi.

See: https://developers.home-assistant.io/docs/development_checklist/
And: https://developers.home-assistant.io/docs/creating_component_code_review/#4-communication-with-devicesservices with link to tutorial how to get started with that.

Olen replied:
Thanks for the review.  First core component I am testing out, so this is a learning process...

RobBie1221 replied:
Great! I'm not sure what the value is of an integration without entities. What kind of entities should the Plant component expose and how would that work without OpenPlantbook as you describe should be possible?

Olen replied:
I agreee that it might look a bit strange, but I could not find a better way to handle this.
The Plant component accepts a lot of max/min values (max_moisture, min_moisture, max_temperature etc.)
https://www.home-assistant.io/integrations/plant/

Currently these have to be set manually for each plant in configuration.yaml.

The OpenPlantbook contains data for all of these settings for a lot of different plant species, so all you need to set for each plant is the name of the species, and the rest can be fetched automatically from the API.  But since the API requires authentication I need a way to configure and store the `client_id` and `secret` in HA to access the API.
I considered doing this directly in the Plant component, but that would require either changes to the way the Plant component is set up to allow setting the `client_id` and `secret` in `configuration.yaml`, or a big rewrite to set up each plant using config_flow, but with so many settings for each plant I think that would be pretty complicated.



So in a follow up PR, I am going to add some more config settings to the Plant component (including `species:`), and also a function that checks if this component (Openplantbook) is set up, and if so - fetch the max/min-values from the API.
The idea is discussed and explained in more detail in this post (and thread):
https://community.home-assistant.io/t/cloud-plant-db-with-api-for-plantcard/225207/2?u=olen

Without Openplantbook, the Plant component will continue to work exactly as today, but with a few more optional config settings.
With Openplantbook, you can skip a lot of the config settings of each plant, and they will be filled in automatically from the API.

RobBie1221 replied:
Hmm...I see your challenge. This requires some architectural discussion I think, let's wait and gather some opinions.

Olen replied:
I'm going to fix the tests etc. in the mean time.

Olen replied:
Btw. I am also slightly confused whether I am allowed to add more config options (e.g `species` and `image`) to the Plant component without converting the whole component to config flow.
As mentioned setting up a `plant` requires quite a lot of config lines with all the sensors etc. which might make a config flow a bit complicated.  And if a config flow is required, it should validate that the configured sensors exist and provides reasonable data etc., so that is pretty heavy.

RobBie1221 replied:
Does it even make sense to reuse Plant integration? Seems to me all that does is group entities and check some min/max values. If you expose entities in this integration with one device per plant it's automatically grouped. Also min/max check isn't that complex to add. You can then taylor the entities here to fit API of OpenPlantbook.

Olen replied:
I'm not sure I understand what you mean.  How would I go ahead and group a bunch of (possibly unrelated) sensors under a new "plantbook-device"?
All the sensors measuring temperature, conductivity etc. are today individual entities without any internal relationship.

Also, "plants" are used in a lot of lovelace cards, automations etc. so I don't want to lose any of that functionality.

RobBie1221 replied:
I may not fully understand the problem. When you use OpenPlantbook, isn't all the data coming from OpenPlantbook? I you create them as entities within this integration you have control over to which device they belong. Or is there no logic when reading from OpenPlantbook what the relation is between measurements?

RobBie1221 replied:
Maybe question would be should Plant use OpenPlantbook or should OpenPlantbook use Plant? I think I lack knowledge of both to effectively discuss this.

Olen replied:
This integration is only to support a simpler configuration of a Plant.
Today, you "must" find the appropriate values and configure this for each plant manually:
```
my_plant:
  max_moisture: 100
  min_moisture: 1
  max_conductivity: 100
  min_conductivity: 1
  max_brightness: 100
  min_brightness: 1
  max_temperature: 100
  min_temperature: 1
  sensors:
    moisture: sensor.xxx
    conductivity: sensor.yyy
    temperature: sensor.zzz
    brightness: sensor.aaa

```
The only thing openplantbook provides are the values for max_* and min_* based on the species of the plant pluss a link to a image of the plant species.

So the next step if/when this integration is accepted is to patch the plant integration to allow for a simplified config for each plant:
```
my_plant:
  species: capsicum annuum
  sensors:
    moisture: sensor.xxx
    conductivity: sensor.yyy
    temperature: sensor.zzz
    brightness: sensor.aaa

```
And when `my_plant` is set up (on startup of HA) all the values for max_* and min_* is fetched from the Openplantbook API and added to the plants config dynamically. The data is only pulled once - on setup of a new plant entity.

So it has nothing to do with the sensors configured for the plant.

RobBie1221 replied:
I start to understand what you are trying to achieve :)

Olen replied:
> I start to understand what you are trying to achieve :)

That is a good first step!  ;)
I added a couple of services to the integration, so it can be used by other integrations to search and get data from the API.
Hopefully that makes it more likely to get added to HA.

Olen replied:
And. btw.  I don't understand how to fix the remaining check that fails:
> codecov/patch — 90.52% of diff hit (target 96.18%)

I am not sure how many other tests I can add that will make any sense.

I am also currently working on documentation of the integration.

balloob committed:
Improve the readability of the config flow code (#42169)

pavoni committed:
Add media browser support to roon media player (#42061)

tim-werner committed:
Rewrite mochad unittest to pytest style (#42164)

alengwenus committed:
Bump pypck to v0.7.4 (#42206)

Shutgun committed:
Handle multiple setups of devolo Home Control (#41752)

emontnemery committed:
Always use shared zeroconf in cast config flow (#42210)

nzapponi committed:
Fix bug for SimpliSafe 2 systems repeatedly saying "your settings have been synchronised" (#42197)

* Fix bug for SimpliSafe 2 systems

* Improved loop and logging for SimpliSafe 2

tkdrob committed:
Correct docstring in Skybell (#42218)

teharris1 committed:
Fix issue in python > 3.8.2 (#42213)

rikroe committed:
Bump bimmer-connected to 0.7.8 to fix login issue (#42215)

Co-authored-by: rikroe <rikroe@users.noreply.github.com>

amelchio committed:
Refresh now() templates on second=0 (#42225)

bramkragten committed:
Update frontend to 20201021.2 (#42242)

emontnemery committed:
Support reloading Tasmota config entries (#42097)

bdraco committed:
Ensure websocket event serializer cache is effective if subscription iden differs (#42226)

Since someone websocket subscriptions will use an iden of 2 for
state_changed event (most comment), and some will use another
number for all events, the cache would not be used because the
iden number was different.  We now cache only the event and
use a fast replace to insert the iden number into the serailized
response.

Adminiuga committed:
Update ZHA dependencies (#42245)

homeassistant committed:
[ci skip] Translation update

bachya committed:
Remove remnants of configurable RainMachine scan interval (#42239)

* Remove remnants of configurable RainMachine scan interval

* Clean up tests

* Cleanup

BrianWithAHat committed:
Fix MaryTTS filename extensions (#42228)

Co-authored-by: Brian Laferriere <BrianWithAHat@users.noreply.github.com>

Olen committed:
Moving API-calls to an external package (pyopenplantbook)

Olen committed:
Moving API-calls to an external package (pyopenplantbook)

tim-werner committed:
Rewrite group unittest to pytest style (#42159)

Olen committed:
Fix tests

Olen committed:
Fix tests

alandtse committed:
Fix is_on test for Tesla Charger switch (#42251)

closes #41755

Olen committed:
Merge branch 'openplantbook' of github.com:Olen/home-assistant into openplantbook

Olen committed:
Add some services

homeassistant commented:
Hello @Olen,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`a6b4f107c36d862488008505689c5e671e95aa43`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;42200&#x2F;commits&#x2F;a6b4f107c36d862488008505689c5e671e95aa43) - This commit has something that looks like an email address (olen@inni.no). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

Olen committed:
Fix service calls

Olen committed:
Add services.yaml

RobBie1221 commented:
I think you messed it up with a rebase, it's showing unrelated changes. You most likely didn't force-pushed your rebase. That will screw up the coverage as well.

Olen commented:
> I think you messed it up with a rebase, it's showing unrelated changes. You most likely didn't force-pushed your rebase. That will screw up the coverage as well.

You are probably right.  Is there a way to undo that, now that I have done a few more commits after that rebase?  My advanced git-foo is not quite up to par...

RobBie1221 commented:
Not sure, maybe quickest way is to open a new PR.

Olen commented:
OK.

Made a completely new branch in my repo and created a new PR: PR#42359

Olen commented:
See https://github.com/home-assistant/core/pull/42359
